jsonrpc/ipc.go

```diff
@@ -0,0 +1,200 @@
+package jsonrpc
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/p2p/netutil"
+)
+
+type Ipc struct {
+	rpc    *Server
+	events NewRequestListener
+	log    utils.SimpleLogger
+
+	connWg sync.WaitGroup // connWg is a WaitGroup for tracking active connections.
+
+	connParams IpcConnParams
+	listener   net.Listener
+
+	// everything below is protected
+	mu      sync.Mutex
+	conns   map[net.Conn]struct{} // conns is a map that holds active connections.
+	closing bool                  // closing indicates whether the server is in the process of shutting down.
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint,
+// and logger. It returns the server or an error if the socket creation fails.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) (*Ipc, error) {
+	unix := &Ipc{
+		rpc:        rpc,
+		log:        log,
+		connParams: DefaultIpcConnParams(),
+		conns:      make(map[net.Conn]struct{}),
+		events:     &SelectiveListener{},
+	}
+	listener, err := createListener(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	unix.listener = listener
+	return unix, nil
+}
+
+// WithConnParams sanity checks and applies the provided params.
+func (i *Ipc) WithConnParams(p *IpcConnParams) *Ipc {
+	i.connParams = *p
+	return i
+}
+
+// WithListener registers a NewRequestListener
+func (i *Ipc) WithListener(listener NewRequestListener) *Ipc {
+	i.events = listener
+	return i
+}
+
+// Run starts the IPC handler in a goroutine and returns any errors encountered during
+// handling. It can be stopped by canceling the provided context or by
```

Comment:
We can un-export `Stop` and require the caller to use the context. Thoughts?

Reply:
Sure.